pm/recipient.go

```diff
@@ -0,0 +1,223 @@
+	ReceiveTicket(ticket *Ticket, sig []byte, seed *big.Int) (won bool, err error)
+
+	// RedeemWinningTicket redeems all winning tickets with the broker
+	// for a session ID
+	RedeemWinningTickets(sessionID string) error
+
+	// TicketParams returns the recipient's currently accepted ticket parameters
+	// for a provided sender ETH adddress
+	TicketParams(sender ethcommon.Address) (*TicketParams, error)
+}
+
+// recipient is an implementation of the Recipient interface that
+// receives tickets and redeems winning tickets
+type recipient struct {
+	val    Validator
+	broker Broker
+	store  TicketStore
+
+	secret [32]byte
+
+	invalidRands sync.Map
+
+	senderNonces     map[string]uint64
+	senderNoncesLock sync.Mutex
+
+	faceValue *big.Int
+	winProb   *big.Int
+}
+
+// NewRecipient creates an instance of a recipient with an
+// automatically generated random secret
+func NewRecipient(broker Broker, val Validator, store TicketStore, faceValue *big.Int, winProb *big.Int) (Recipient, error) {
+	randBytes := make([]byte, 32)
+	if _, err := rand.Read(randBytes); err != nil {
+		return nil, err
+	}
+
+	var secret [32]byte
+	copy(secret[:], randBytes[:32])
+
+	return NewRecipientWithSecret(broker, val, store, secret, faceValue, winProb), nil
+}
+
+// NewRecipientWithSecret creates an instance of a recipient with a user provided
+// secret. In most cases, NewRecipient should be used instead which will
+// automatically generate a random secret
+func NewRecipientWithSecret(broker Broker, val Validator, store TicketStore, secret [32]byte, faceValue *big.Int, winProb *big.Int) Recipient {
+	return &recipient{
+		broker:       broker,
+		val:          val,
+		store:        store,
+		secret:       secret,
+		faceValue:    faceValue,
+		senderNonces: make(map[string]uint64),
+		winProb:      winProb,
+	}
+}
+
+// ReceiveTicket validates and processes a received ticket
+func (r *recipient) ReceiveTicket(ticket *Ticket, sig []byte, seed *big.Int) (bool, error) {
+	recipientRand := r.rand(seed, ticket.Sender)
+
+	if crypto.Keccak256Hash(ethcommon.LeftPadBytes(recipientRand.Bytes(), uint256Size)) != ticket.RecipientRandHash {
+		return false, errors.Errorf("invalid recipientRand generated from seed %v", seed)
+	}
+
+	if ticket.FaceValue.Cmp(r.faceValue) != 0 {
+		return false, errors.Errorf("invalid ticket faceValue %v", ticket.FaceValue)
+	}
+
+	if ticket.WinProb.Cmp(r.winProb) != 0 {
+		return false, errors.Errorf("invalid ticket winProb %v", ticket.WinProb)
+	}
+
+	if err := r.val.ValidateTicket(ticket, sig, recipientRand); err != nil {
+		return false, err
+	}
+
+	if !r.validRand(recipientRand) {
+		return false, errors.Errorf("invalid already revealed recipientRand %v", recipientRand)
+	}
+
+	if err := r.updateSenderNonce(recipientRand, ticket.SenderNonce); err != nil {
+		return false, err
+	}
+
+	if r.val.IsWinningTicket(ticket, sig, recipientRand) {
+		if err := r.store.Store(ticket.RecipientRandHash.Hex(), ticket, sig, recipientRand); err != nil {
```

Comment:
What do you imagine this helper to look like?

At the moment, the sender generates a sessionID in `StartSession()` using the provided `TicketParams` struct:

```
sessionID := ticketParams.RecipientRandHash.Hex()
```

Meanwhile, the recipient generates a sessionID in `ReceiveTicket()` using the provided `Ticket` struct:

```
sessionID := ticket.RecipientRandHash.Hex()
```

Since the sender and recipient use different data structures in order to access `RecipientRandHash`, the common input for both of them would just be `RecipientRandHash` and at that point the sessionID helper would just return the hex string representation of the input as the output.